### PR TITLE
[7/17] Fix periods module doctests

### PR DESCRIPTION
**Pull request by maukoquiroga**
Part of #1061 

Thanks for contributing to OpenFisca! Remove this line, as well as any other in the following that don't fit your contribution  :)

#### Technical changes

- Fix `openfisca_core.periods` doctests.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -51,5 +51,8 @@ non_interactive     = True
 [mypy-openfisca_core.commons.tests.*]
 ignore_errors       = True
 
+[mypy-openfisca_core.periods.tests.*]
+ignore_errors 		= True
```
Comment (MattiSG) [inline setup.cfg:55]: What do we get from both adding this file to the list of tests and ignoring the resulting errors? 🤔

File: openfisca_tasks/test_code.mk [hunk 2]
```diff
@@ -20,7 +20,7 @@ test-code: test-core test-country test-extension
 	@$(call print_pass,$@:)
 
 ## Run openfisca-core tests.
-test-core: $(shell pytest --quiet --quiet --collect-only | cut -f 1 -d ":")
+test-core: $(shell pytest --quiet --quiet --collect-only 2> /dev/null | cut -f 1 -d ":")
```
Comment (MattiSG) [inline openfisca_tasks/test_code.mk:23]: > While I understand from the commit message that this change aims to “Ignore test collection prior install”, I am a bit concerned that we never see errors if they were to happen, making it very hard to debug.

**Review by MattiSG — COMMENTED** [review]
Thank you for these contributions! I still don't understand how to validate doctests, and I suffered a lot from trying to set them straight (cf. https://github.com/openfisca/openfisca-core/pull/1134). I see here that we in the end ignore errors for the “periods” module. If that is indeed the case, then could you help me understand what is the point of “fixing” those tests? 🙂

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,18 @@
 # Changelog
 
+### 35.8.6 [#1138](https://github.com/openfisca/openfisca-core/pull/1138)
```
[the remark this answers is not in the document]
Comment (MattiSG) [reply CHANGELOG.md:3]: Okay, this depends on #1137.

File: openfisca_tasks/test_code.mk (hunk 2 above)
Comment (MattiSG) [reply openfisca_tasks/test_code.mk:23]: This is actually part of #1137.

**Comment by MattiSG** [comment]
In order to ease future reviews, when you open a PR that depends on another one, could you please make it explicit in the description? 🙂 This will make it easier to review 😊

File: setup.cfg (hunk 1 above)
Comment (maukoquiroga) [reply setup.cfg:55]: This line in particular is to ignore enforcing type checks in tests. While I do believe as we've discussed before type hints are great code documentation, I also think there's a time for them: maturity. Proper testing is not this library's best feature, thus the line above. This hypothesis can of course be challenged.

File: openfisca_tasks/test_code.mk (hunk 2 above)
Comment (maukoquiroga) [reply openfisca_tasks/test_code.mk:23]: Yes, sorry.

File: openfisca_core/periods/instant_.py [hunk 4]
```diff
@@ -1,183 +1,165 @@
...
-
-        """
         instant_str = config.str_by_instant_cache.get(self)
         if instant_str is None:
             config.str_by_instant_cache[self] = instant_str = self.date.isoformat()
         return instant_str
 
     @property
     def date(self):
-        """
```
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/instant_.py:42]: We don't: these are _properties_, closer to attributes than methods. They're now documented in the class' doc.

**Comment by maukoquiroga** [comment]
> Thank you for these contributions! I still don't understand how to validate doctests, and I suffered a lot from trying to set them straight (cf. #1134). I see here that we in the end ignore errors for the “periods” module. If that is indeed the case, then could you help me understand what is the point of “fixing” those tests? 🙂

Hi. Actually this PR make periods' doctest run (that's how I fixed them), so for them to be part of the test suite. What is ignored, at least for now, are tests' type checks, that's it. 

Doctests are valuable in that they're the rare expression of unit tests in this library, coming from the contributors who gave birth to much the codebase.

File: openfisca_core/periods/period_.py [hunk 5]
```diff
@@ -183,168 +264,200 @@ def get_subperiods(self, unit):
...
 
     @property
     def size_in_months(self):
-        """
-        Return the size of the period in months.
+        """Return the size of the period in months."""
 
-        >>> period('month', '2012-2-29', 4).size_in_months
-        4
-        >>> period('year', '2012', 1).size_in_months
```
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:345]: Because, as a property, just one is enough, and it can be found in the class' doc, as per good practices (you can of course always document properties, but the syntax is different from normal methods and functions, and from experience is extremely painful to get sphinx to generate them usefully).

I concluded then that a) these are closer to attributes, and just one doctest is enough, or they are methods, in which case they should be declared so, and make a breaking change.

In an ideal world, we would have these properties calculated and cached as any attribute. The only reason we don't is because these classes (period and instance) inherit from _tuple_, which imposes these limitations.

This hack of inheriting from tuples was done back then for performance, in Python 2.x. Today, we've better solutions in the standard library (for example dataclasses with immutable properties).

Furthermore, I think random declarative tests without edge-case of any math logic are useless waste of CPU.

[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:345]: In short, before this PR, no doctest was run. So no test was removed, actually one was added 😁

[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:345]: Reflecting on my previous answer, your question made me question the actual value of doctests, as a general practice, and in particular in OpenFisca:

_Doctests are useful as examples and to ensure documentation is up-to-date. They're not exhaustive as in most cases that would render the documentation harder to read. Exhaustive testing is done with unit tests._

If you agree to that previous statement, then what's missing are unit tests.

[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:345]: Done!

**Comment by maukoquiroga** [comment]
Hello @MattiSG ! I hope to have addressed all of your concerns. I have also updated the description to make it clear that this PR only addresses "tests", including those in the docstrings (doctests).

File: openfisca_core/periods/instant_.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (MattiSG) [reply openfisca_core/periods/instant_.py:42]: Thank you for this clarification! Would you have a reference to documentation on how this works? I currently have no understanding of the mechanism at all.

File: openfisca_core/periods/period_.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MattiSG) [reply openfisca_core/periods/period_.py:345]: Thank you for having clarified this point!

I cannot know if I agree with that statement as I still don't understand how doctests are compiled, run and exploited.
What I see is that they are incredibly painful to write and maintain, that barely any contributor seems able to maintain them, and that on the other hand we have a reliable and standard system for tests. As it is, I am indeed tempted of discarding all “doctests” since no matter how much work is put in (and you do put a lot of effort in it), it seems we end up disabling them because they don't pass or they have too many side effects.

As already suggested elsewhere, I believe it would be good that we have a wider discussion on the testing strategy.

File: setup.cfg (hunk 1 above)
Comment (MattiSG) [reply setup.cfg:55]: That sounds legitimate 🙂 I would welcome a comment on it though.

**Review by MattiSG — COMMENTED** [review]
I cannot review this as it depends on #1141 that is disputed, and I don't support changing so many dependencies in an untestable way (cf. https://github.com/openfisca/openfisca-core/issues/1140#issuecomment-1200902579).

In order to ease reviews, could you please avoid force pushing once the PR has been opened, and rely instead on `fixup` commits if necessary? 🙂 Otherwise, I cannot rely on GitHub’s presentation of what has “changed since last review”, forcing me to spend time reviewing the same changeset again.

File: openfisca_core/periods/period_.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:345]: Doctests, that is the examples part of the doc, are injected during the test collection and executed as normal tests.

As such, as you guessed, this PR is more about tests than doc.

My intent here is to reuse them as much as possible to add actual unit tests, removing them from the doc so as to make the doc less cluttered, and just leave one or two for informative purposes, keeping the doc congruent with the evolution of the API.

Please do make a difference between doctests, that is the "examples" part of the doc, which are normal tests, and other stuff like typing and documentation parsing, which are not covered here.

For example what's being disabled here are not doctests, but type hints for the /tests folder.

My "quest" is to transform as much as we can those doctests into unit tests :)

**Comment by maukoquiroga** [comment]
@MattiSG I've had cleanup as much as I can. 

IMHO both readability and test coverage have improved, which is a triple win:

1. We know several functions to deprecate soon.
2. We can now more safely extend periods! :) (I've found some bugs downstream already).
3. I forgot this one (something related to types, contracts, and maintainability).

**Review by benjello — COMMENTED** [review]
It might be worthwhile to run openfisca-france using this PR to fix here and there the impacts on a  large "prod" repo.

**Comment by maukoquiroga** [comment]
Hmm I did a `git mv` but it still marked it as deleted.

**Comment by maukoquiroga** [comment]
> It might be worthwhile to run openfisca-france using this PR to fix here and there the impacts on a large "prod" repo.

@benjello This pull request doesn't add any feature but deprecates some and fixes several bugs. In general, it renders the whole module more reliable, as many edge-cases that went silent before will now yield a useful error. Yet, of course, that means breaking changes as the range of the acceptable shrunk. It's a fair trade-off and a much needed contribution, but requires thorough testing. I'd be ideal to run this against France with population data :)

**Comment by benjello** [comment]
@maukoquiroga : as is it breaks openfisca-france test because `period` method is deprecated and replaced by `build_period` so it might be efficient to migrate the changes in openfisca-france concomitantly.

What do you think @sandcha @MattiSG ?

**Comment by maukoquiroga** [comment]
> @maukoquiroga : as is it breaks openfisca-france test because `period` method is deprecated and replaced by `build_period` so it might be efficient to migrate the changes in openfisca-france concomitantly.
> 
> 
> 
> What do you think @sandcha @MattiSG ?  

The rename is a find & replace, I'm more wondering if the stringer contracts will break parts of the model. For example, some cases returning None now fail, invalid periods like 2022-1-1 also fail, etc. 

Would this be suitable @MattiSG for your proposal of a release candidate?

**Comment by MattiSG** [comment]
> Would this be suitable @MattiSG for your proposal of a release candidate?

Yes, of course, this is a good case for #1159. But I want to stress once more that #1159 is **not** about _allowing_ the use of release candidates, but about allowing replacing code review by extensive testing in some contexts. In the given case, I understand that @benjello says testing on France would be welcome. If you want to simplify that testing by using a release candidate (as opposed to a local clone), sure, do it! No need for #1159 for that 😉

File: openfisca_core/periods/period_.py [hunk 6]
```diff
@@ -1,489 +1,588 @@
...
+
+        if unit == MONTH and self.unit == YEAR:
+            return self.size * 12
+
+        raise ValueError(
+            f"Cannot calculate number of {unit.plural} in a "
+            f"{str(self.unit)}."
+            )
+
+    def this(self, unit: DateUnit) -> Period:
```
Comment (maukoquiroga) [inline openfisca_core/periods/period_.py:338]: @benjello look from here below, what do you think?

**Comment by maukoquiroga** [comment]
@eraviart What do you think?

File: openfisca_core/periods/period_.py (hunk 6 above)
Comment (maukoquiroga) [reply openfisca_core/periods/period_.py:338]: @eraviart
